DSharpPlus.VoiceNext/Interop/Bindings.cs

```diff
@@ -0,0 +1,156 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+using System;
+using System.Runtime.InteropServices;
+
+namespace DSharpPlus.VoiceNext.Interop
+{
+    internal static unsafe class Bindings
+    {
+        [DllImport("libopus", CallingConvention = CallingConvention.Cdecl)]
+        private static extern IntPtr opus_encoder_create(int samplingRate, int channels, int application, out OpusError error);
+
+        [DllImport("libopus", CallingConvention = CallingConvention.Cdecl)]
+        private static extern void opus_encoder_destroy(IntPtr encoder);
+
+        [DllImport("libopus", CallingConvention = CallingConvention.Cdecl)]
+        private static extern int opus_encode(IntPtr encoder, byte* pcm, int frameSize, byte* data, int maxDataBytes);
+
+        [DllImport("libopus", CallingConvention = CallingConvention.Cdecl)]
+        private static extern OpusError opus_encoder_ctl(IntPtr encoder, OpusControl ctl, int value);
+
+        [DllImport("libopus", CallingConvention = CallingConvention.Cdecl)]
+        private static extern IntPtr opus_decoder_create(int sampleRate, int channels, out OpusError error);
+
+        [DllImport("libopus", CallingConvention = CallingConvention.Cdecl)]
+        private static extern void opus_decoder_destroy(IntPtr decoder);
+
+        [DllImport("libopus", CallingConvention = CallingConvention.Cdecl)]
+        private static extern int opus_decode(IntPtr decoder, byte* opusData, int opusDataLength, byte* data, int frameSize, int decodeFec);
+
+        [DllImport("libopus", CallingConvention = CallingConvention.Cdecl)]
+        private static extern int opus_packet_get_nb_channels(byte* data);
+
+        [DllImport("libopus", CallingConvention = CallingConvention.Cdecl)]
+        private static extern int opus_packet_get_nb_frames(byte* data, int length);
+
+        [DllImport("libopus", CallingConvention = CallingConvention.Cdecl)]
+        private static extern int opus_packet_get_samples_per_frame(byte* data, int samplingRate);
+
+        [DllImport("libopus", CallingConvention = CallingConvention.Cdecl)]
+        private static extern int opus_decoder_ctl(IntPtr decoder, OpusControl ctl, out int value);
+
+        public static IntPtr CreateEncoder(int sampleRate, int channelCount, int application)
+        {
+            var encoder = opus_encoder_create(sampleRate, channelCount, application, out var error);
+            return error == OpusError.Ok ? encoder : throw new Exception($"Failed to instantiate Opus encoder: {error} ({(int)error})");
```

Comment:
Why are we casting OpusError to an int

Reply:
Nevermind, I failed to read

Reply:
to print the raw error code, in case a future libopus version adds a new error code we dont have represented in `OpusError`

Reply:
Would that throw an exception trying to print `error` without an int cast then

Reply:
iirc that varies between netfx/mono/core (vague memory)